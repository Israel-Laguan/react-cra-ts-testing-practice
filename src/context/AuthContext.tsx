/* eslint-disable */
import React, { createContext, useState, useContext } from 'react'

const AuthContext = createContext<{
  isAuthenticated?: boolean
  setIsAuthenticated?: Function
}>({})

interface AuthProviderProps {
  children: React.ReactNode
}
export const AuthProvider = ({ children }: AuthProviderProps) => {
  const [isAuthenticated, setIsAuthenticated] = useState(false)
  const { Provider } = AuthContext

  {
    setTimeout(() => {
      setIsAuthenticated(true)
    }, 3000)
  }

  return <Provider value={{ isAuthenticated }}>{children}</Provider>
}

export const useAuthContext = () => useContext(AuthContext)
