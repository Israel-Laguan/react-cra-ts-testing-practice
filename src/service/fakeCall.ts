export const fakeCall = (message: string): Promise<string> =>
  new Promise((resolve) => {
    setTimeout(() => {
      resolve(message)
    }, 4000)
  })
