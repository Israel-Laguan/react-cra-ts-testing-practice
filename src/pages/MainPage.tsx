import { useState, useEffect } from 'react'

import { fakeCall } from '../service/fakeCall'
import { useAuthContext } from '../context/AuthContext'

const MainPage = () => {
  const [result, setResult] = useState('')
  const [isLoading, setIsLoading] = useState(true)
  const { isAuthenticated } = useAuthContext()

  const changeResult = async (newMessage: string) => {
    setResult(await fakeCall(newMessage))
  }

  useEffect(() => {
    ;(async () => {
      const result: string = await fakeCall('DONE!')
      setResult(result)
      setIsLoading(false)
    })()
  }, [])
  return isLoading ? (
    <div className="danger" data-testid="Div::Loading">
      Loading...
    </div>
  ) : (
    <>
      <div className="success" data-testid="Div::Result">
        {result}
      </div>
      <div>{isAuthenticated ? 'Welcome!' : 'You Are Not Logged In!'}</div>
      <div>
        {isAuthenticated && (
          <button
            onClick={(e) => {
              e.preventDefault()
              changeResult('Result has changed!')
            }}
          >
            Change Result
          </button>
        )}
      </div>
    </>
  )
}
export default MainPage
