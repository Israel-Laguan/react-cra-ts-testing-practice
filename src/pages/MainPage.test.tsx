/* eslint-disable */
import MainPage from './MainPage'
import { useAuthContext, AuthProvider } from '../context/AuthContext'

import { fakeCall } from '../service/fakeCall'
import { render, act } from '@testing-library/react'
import React from 'react'

jest.mock('../service/fakeCall', () => ({
  fakeCall: jest.fn(),
}))

jest.mock('../context/AuthContext', () => ({
  ...jest.requireActual('../context/AuthContext'),
  AuthProvider: ({ children }: { children: React.ReactNode }) => (
    <>{children}</>
  ),
  useAuthContext: jest.fn(),
}))

describe('Test Main Page', () => {
  it('Match snapshot', () => {
    const { asFragment } = render(<MainPage />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('Verify if text exists', () => {
    const { getByText } = render(<MainPage />)
    const text = getByText(/Loading.../)
    expect(text).toBeInTheDocument()
  })

  it('Verify if data test id exists', () => {
    const { getByTestId } = render(<MainPage />)
    const element = getByTestId(/Div::Loading/)
    expect(element).toBeInTheDocument()
  })

  it('Verify if loaded', () => {
    const { getByTestId } = render(<MainPage />)
    const element = getByTestId(/Div::Result/)
    expect(element).toBeInTheDocument()
  })

  it('Verify if authenticated', () => {
    const { getByText } = render(<MainPage />)
    const element = getByText(/Welcome/)
    expect(element).toBeInTheDocument()
  })

  it('Verify call fetch result', () => {
    const { getByText } = render(<MainPage />)
    const element = getByText(/DONE!/)
    expect(element).toBeInTheDocument()
  })

  it('Wait state change', async () => {
    useAuthContext.mockReturnValue({
      isAuthenticated: false,
    })
    let component: any
    await act(async () => {
      component = render(<MainPage />)
    })
    const element = component.getByTestId(/Div::Result/)
    expect(element).toBeInTheDocument()
  })

  it('Context value', async () => {
    useAuthContext.mockReturnValue({
      isAuthenticated: true,
    })
    let component: any
    await act(async () => {
      component = render(<MainPage />)
    })
    const element = component.getByText(/Welcome!/)
    expect(element).toBeInTheDocument()
  })

  it('Async promises resolve', async () => {
    useAuthContext.mockReturnValue({
      isAuthenticated: true,
    })
    fakeCall.mockResolvedValue('TEST!')

    let component: any
    await act(async () => {
      component = render(<MainPage />)
    })
    const element = component.getByText(/TEST!/)
    expect(element).toBeInTheDocument()
  })
})
