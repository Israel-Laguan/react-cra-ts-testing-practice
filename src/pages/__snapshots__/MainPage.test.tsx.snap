// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Main Page Match snapshot 1`] = `
<DocumentFragment>
  <div
    class="danger"
    data-testid="Div::Loading"
  >
    Loading...
  </div>
</DocumentFragment>
`;
